### ISSUE TITLE
Inconsistencies destructuring a tuple type with an optional element

### ISSUE THREAD

**Issue by tmkn** [microsoft/TypeScript#31770]
**TypeScript Version:**  3.4.0-dev.201xxxxx

**Search Terms:**

**Code**
With `strictNullChecks` enabled, TypeScript has problems propagating an optional tuple element through a destructuring assignment.
```ts
type KeyValue = [string, string?];

let [key, value]: KeyValue = ["foo"];
value.toUpperCase(); //No error, falsely inferred as string
```
The following works though:
```ts
let keyValue: KeyValue = ["foo"];
keyValue[1].toUpperCase(); //Error, Object is possibly 'undefined'.

let [key2, value2] = keyValue;
value2.toUpperCase(); //Error, Object is possibly 'undefined'.

let [key3, value3]: KeyValue = ["foo"] as KeyValue;
value3.toUpperCase(); //Error, Object is possibly 'undefined'.
```
**Expected behavior:**
```ts
type KeyValue = [string, string?];

let [key, value]: KeyValue = ["foo"];
value.toUpperCase();  //Error, Object is possibly 'undefined'.
```

**Actual behavior:**
```ts
type KeyValue = [string, string?];

let [key, value]: KeyValue = ["foo"];
value.toUpperCase();  //No error is raised
```

**Playground Link:** 
http://www.typescriptlang.org/play/index.html#src=type%20KeyValue%20%3D%20%5Bstring%2C%20string%3F%5D%3B%0A%0Alet%20%5Bkey%2C%20value%5D%3A%20KeyValue%20%3D%20%5B%22foo%22%5D%3B%0Avalue.toUpperCase()%3B%20%2F%2FNo%20error%2C%20falsely%20inferred%20as%20string%0A%0Alet%20keyValue%3A%20KeyValue%20%3D%20%5B%22foo%22%5D%3B%0AkeyValue%5B1%5D.toUpperCase()%3B%20%2F%2FError%2C%20Object%20is%20possibly%20'undefined'.%0A%0Alet%20%5Bkey2%2C%20value2%5D%20%3D%20keyValue%3B%0Avalue2.toUpperCase()%3B%20%2F%2FError%2C%20Object%20is%20possibly%20'undefined'.%0A%0Alet%20%5Bkey3%2C%20value3%5D%3A%20KeyValue%20%3D%20%5B%22foo%22%5D%20as%20KeyValue%3B%0Avalue3.toUpperCase()%3B%20%2F%2FError%2C%20Object%20is%20possibly%20'undefined'.
**Related Issues:**

**Comment by ajafff** [comment]
I actually expected an error at the type annotation of that declaration:

```ts
let [key, value]: KeyValue = ["foo"];
//              ~~~~~~~~~~ where is there no error?
```

**Comment by fatcerberus** [comment]
I assume there's no error for the destructuring because it's still a 2-tuple, just that the second element is optional so we can treat it as `string | undefined` (and in fact TypeScript does exactly that).  Tuples with optional elements would be much less useful (IMO) if you couldn't actually use destructuring to access the optional elements.  `let [ key, value? ]: KeyValue = ...` isn't legal syntax so this is the best TS can do, methinks.

That being said, the behavior here is weird:
```ts
// hover over 'key' - type is string
// hover over 'value' - type is string | undefined
let [ key, value ]: KeyValue = [ "foo" ];

// hover over 'key' and 'value' in turn - type of both is now string
key.toUpperCase();
value.toUpperCase();
```

So TS knows `value` could be undefined at the point of assignment, but then promptly forgets and prematurely narrows it to string.

**Comment by fatcerberus** [comment]
I think the underlying cause is that `[ "foo" ]` is probably inferred as an array `string[]` rather than a 1-tuple `[ string ]`; it’s well-documented in this issue tracker that TS prefers to infer arrays over tuples.  That’s why the `as KeyValue` cast fixes it.

Could be fixed with smarter contextual typing, methinks.

**Comment by tmkn** [comment]
To add, destructured function arguments work
```ts
function test1([key, value]: KeyValue) {
    value.toUpperCase(); //Error, Object is possibly 'undefined
}

//even with a default parameter
function test2([key, value]: KeyValue = ["foo"]) {
    value.toUpperCase(); //Error, Object is possibly 'undefined
}
```

**Comment by tmkn** [comment]
It seems it breaks if the union type involves `string`:

It correctly raises an error if the optional tuple element is not a `string` e.g. a `number`
```ts
{
    type KeyValue = [string, number?];
    let [key, value]: KeyValue = ["foo"];
    value.toString(); //Error, Object is possibly 'undefined
}
```
It doesn't raise an error if the optional tuple element is a union of `string` and `number` 
Actually it is inferred as `string` o_O
```ts
{
    type KeyValue = [string, (number | string)?];
    let [key, value]: KeyValue = ["foo"];
    value.toString(); //No error, even though a union, inferred as string
    value.toUpperCase(); //fine
    value.toFixed(2);//Property 'toFixed' does not exist on type 'string'. Did you mean 'fixed'?
}
```
It however behaves correctly if the union **doesn't** involve a string
```ts
{
    type KeyValue = [string, (number | boolean)?];
    let [key, value]: KeyValue = ["foo"];
    value.toString(); //Error, Object is possibly 'undefined

    if (typeof value !== "undefined") {
        //value inferred as number | boolean
    }
}
```
And it's broken again if we add `string` to the union
Again it is inferred as `string` even though it's a union type.
```ts
{
    type KeyValue = [string, (number | boolean | string)?];
    let [key, value]: KeyValue = ["foo"];
    value.toString(); ////No error, falsely inferred as string
}
```
Even if we are being explicit and add `undefined` to the union, it still doesn't warn about the object being possibly `undefined`
Again it is falsely inferred as `string`
```ts
{
    type KeyValue = [string, (number | string | undefined)?];
    let [key, value]: KeyValue = ["foo"];
    value.toString(); //No error, falsely inferred as string
}
```

**Comment by fatcerberus** [comment]
Yeah, it’s all because the array literal `["foo"]` is inferred as array `string[]` rather than tuple (i.e. `typeof ["foo"][1] == string`.  `string[]` obviously doesn’t work as input to a tuple that requires a `number` element; making it a union including `string` makes that work again.

`["foo"] as const` should also fix it by coercing the array literal to a `readonly` tuple.

Like I said, I think this could be made to be an error with some better contextual typing.  Right now I’m thinking it might be a design limitation.

**Comment by fatcerberus** [comment]
@tmkn Basically this is what the TypeScript compiler sees:

```ts
type KeyValue = [string, string?];

let tmp: string[] = [ "foo" ];

// 'typeof tmp[1]' is string so 'value' is now also string
let [ key, value ]: KeyValue = tmp;
```
